### DISCUSSION TITLE
Best practice for a "global" static component on all static pages?

### DISCUSSION THREAD

**Discussion by sessa** [vercel/next.js#12227] [Help]
I am trying to figure out the best approach to include a navigation that requires an API call to be rendered that exists on all pages and still be able to SSG.  I see two ways of accomplishing as of now:

1. Include the API call in the `getStaticProps` export in each page.
2. Include the API call in `getInitialProps` in `_app.tsx` and then include an empty `getStaticProps` export in each static page, triggering SSG.

```typescript
export const getStaticProps: GetStaticProps = async () => {
  return {
    props: {},
  }
}
```

The second option is the cleanest as of now, but I was wondering if there is a way to accomplish this "globally".  Ideally, including `getStaticProps` in `_app.tsx` is how I'd imagine it could work?

**Comment by sessa** [comment]
A quick update:

It seems like this solution seem to match the #2 option I provided above: [#10949 - comment](10949#discussioncomment-6884)
> @fabb So, looking at your workaround 👆, your hasStaticProps is "useless", it should generate a SSG page even if you remove it. (but it'll will run the getInitialProps for nothing, which may not be what you want).
An alternative way to do that would be to check for the existence of getStaticProps instead of hasStaticProps to avoid using a useless variable that may not be "in-sync" with the actual use of getStaticProps. By doing so, you'd just have to add getStaticProps to all SSG pages (even if you don't fetch anything) and you could still use _app getStaticProps and it'll build static pages properly.

If this is the only current solution, I can close this discussion out.

**Comment by valse** [comment]
Hi, but in the [doc](https://nextjs.org/docs/advanced-features/custom-app) there is a note:

> Adding a custom getInitialProps in your App will disable Automatic Static Optimization.

With the second option this is true or the common server request is resolved at build time rendering a SSG page?
